core/trino-main/src/main/java/io/trino/sql/planner/rowpattern/LogicalIndexPointer.java

```diff
@@ -0,0 +1,122 @@
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static java.util.Objects.requireNonNull;
+
+public class LogicalIndexPointer
+{
+    // logical position is a position among rows tagged with certain label (or label from a certain set)
+    // it has the following semantics:
+    // start from FIRST or LAST row tagged with the label (with RUNNING or FINAL semantics), and go logicalOffset steps forward (for FIRST) or backward (for LAST),
+    // skipping to consecutive rows with matching label
+    // Default: RUNNING LAST offset = 0
+    private final Set<IrLabel> label;
+    private final boolean last;
+    private final boolean running;
+    private final int logicalOffset;
+
+    // physical offset is the offset in physical rows, starting from the logical position. negative for PREV, positive for NEXT. The default is -1 for PREV and 1 for NEXT.
+    // Unspecified physical offset defaults to 0.
+    private final int physicalOffset;
+
+    @JsonCreator
+    public LogicalIndexPointer(Set<IrLabel> label, boolean last, boolean running, int logicalOffset, int physicalOffset)
+    {
+        this.label = requireNonNull(label, "label is null");
+        this.last = last;
+        this.running = running;
+        checkArgument(logicalOffset >= 0, "logical offset must be >= 0, actual: " + logicalOffset);
+        this.logicalOffset = logicalOffset;
+        this.physicalOffset = physicalOffset;
+    }
+
+    @JsonProperty
+    public Set<IrLabel> getLabel()
+    {
+        return label;
+    }
+
+    @JsonProperty
+    public boolean isLast()
+    {
+        return last;
+    }
+
+    @JsonProperty
+    public boolean isRunning()
+    {
+        return running;
+    }
+
+    @JsonProperty
+    public int getLogicalOffset()
+    {
+        return logicalOffset;
+    }
+
+    @JsonProperty
+    public int getPhysicalOffset()
+    {
+        return physicalOffset;
+    }
+
+    public LogicalIndexNavigation toLogicalIndexNavigation(Map<IrLabel, Integer> mapping)
+    {
+        return new LogicalIndexNavigation(
+                label.stream()
+                        .map(mapping::get)
+                        .collect(toImmutableSet()),
+                last,
+                running,
+                logicalOffset,
+                physicalOffset);
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        LogicalIndexPointer that = (LogicalIndexPointer) o;
+        return last == that.last &&
+                running == that.running &&
+                logicalOffset == that.logicalOffset &&
+                physicalOffset == that.physicalOffset &&
+                label.equals(that.label);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(label, last, running, logicalOffset, physicalOffset);
+    }
+}
```

Comment:
No, it's not a map key or so. I added equals and hashCode because I thought of future optimization which could identify duplicate pointers. Should I remove these methods for now?

Reply:
Yes, let's get rid of them for now

Reply:
Actually, equals is used by the serialization test. I restored the methods.